[WGSL] Add more type declarations

tadeuzagallo opened the pull request:
#### 89a5ffd4e05fbe0e493c1266c8ad358172322837
<pre>
[WGSL] Add more type declarations
<a href="https://bugs.webkit.org/show_bug.cgi?id=255115">https://bugs.webkit.org/show_bug.cgi?id=255115</a>
rdar://107726330

Reviewed by Dean Jackson.

Add declarations for binary minus, multiplication of vectors and vector
constructors from a single element (which was apparently missing from
the spec)

* Source/WebGPU/WGSL/TypeDeclarations.rb:

Canonical link: <a href="https://commits.webkit.org/262836@main">https://commits.webkit.org/262836@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3ac1c19748383e1c7d234f2c2d44310906f465f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2753 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2752 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2899 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4163 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3114 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2850 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2420 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2779 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3211 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2461 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3932 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/705 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2446 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2302 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2462 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2497 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3680 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2832 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2273 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2494 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2449 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/677 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2487 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2646 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WGSL/TypeDeclarations.rb:
```diff
@@ -8,13 +8,22 @@
 }
 
 operator :*, {
+    [T < Number, N].(Vector[T, N], T) => Vector[T, N],
```
nit: I wonder if we should add comments to document what these overloads are for? Eg:

```
# Vector scaling

djg commented on Source/WebGPU/WGSL/TypeDeclarations.rb:
```diff
@@ -8,13 +8,22 @@
+    [T < Number, N].(Vector[T, N], T) => Vector[T, N],
+    [T < Number, N].(T, Vector[T, N]) => Vector[T, N],
+    [T < Number, N].(Vector[T, N], Vector[T, N]) => Vector[T, N],
     [T < Float, C, R].(Matrix[T, C, R], Vector[T, C]) => Vector[T, R],
```
```
# Matrix-vector multiplication
```

djg commented on Source/WebGPU/WGSL/TypeDeclarations.rb:
```diff
@@ -8,13 +8,22 @@
 operator :*, {
+    [T < Number, N].(Vector[T, N], T) => Vector[T, N],
+    [T < Number, N].(T, Vector[T, N]) => Vector[T, N],
+    [T < Number, N].(Vector[T, N], Vector[T, N]) => Vector[T, N],
```
aside: I checked the spec, but component-wise multiplication of vectors seems like a strange operation.

djg reviewed (approved):
:shipit:

tadeuzagallo committed:
[WGSL] Add more type declarations
https://bugs.webkit.org/show_bug.cgi?id=255115
rdar://107726330

Reviewed by Dean Jackson.

Add declarations for binary minus, multiplication of vectors and vector
constructors from a single element (which was apparently missing from
the spec)

* Source/WebGPU/WGSL/TypeDeclarations.rb:

Canonical link: https://commits.webkit.org/262836@main
